### PR TITLE
require subcommand for all actions

### PR DESCRIPTION
**Pull request by xmatthias**
## Summary
Merging #2264 into develop.

Develop has been merged into that branch - and helpstrings have been adjusted to account for changes from develop.

## changes 
(mostly copied from #2264)
This PR tries to address this as follows:

* introduce `freqtrade trade` subcommand for trading in dry / live runs
* add --dry-run to force dry-runs
* Remove usage of `freqtrade ` without subcommand
* only accept arguments after subcommands (so it's aligned throughout the interface)
* use grouped parsers to group common arguments together (for visuals / help printing).
* removal of Defaultstrategy as default for `--strategy`
* removal of DefaultHyperopt as default for `--hyperopt`

### CODE REVIEW

**Comment by hroff-1902** [comment]
I'm amused, it's merged with develop after list-pairs/markets with only one small conflict...

**Comment by xmatthias** [comment]
fixing the test-failure is easy...

File: freqtrade/configuration/arguments.py [hunk 1]
```diff
@@ -110,87 +101,122 @@ def _build_subcommands(self) -> None:
...
         )
         list_timeframes_cmd.set_defaults(func=start_list_timeframes)
         self._build_args(optionlist=ARGS_LIST_TIMEFRAMES, parser=list_timeframes_cmd)
 
         # Add list-markets subcommand
         list_markets_cmd = subparsers.add_parser(
             'list-markets',
-            help='Print markets on exchange.'
+            help='Print markets on exchange.',
+            parents=[_common_parser],
```
Comment (xmatthias) [inline freqtrade/configuration/arguments.py]: While i've added this now to fix the broken tests, i'm not sure if including the parent parser makes sense.

I did intentionally not do this for `create-userdir` - so it does not accept `--config` as argument - which means that no config file is not loaded (only the basic one), even if `config.json` does exist, fixing some oddities with these commands.

I'll leave the decision to you if this makes sense for these 2 commands ...

**Comment by hroff-1902** [comment]
Let's postpone merging this into develop for several days/a week. I'd like to try using it as I 'normally' use freqtrade, not only in the development mode...

**Comment by hroff-1902** [comment]
^^ Strange merge...

**Comment by xmatthias** [comment]
why? by default you're only seeing conflicts - there is a link at the bottom:
```
You are viewing a condensed version of this merge commit. You can view the full changes here.
```

Note: the merge is not related to the failure, but removal of POWR/BTC from bittrex is ... 
will have a look at that tomorrow ...

File: freqtrade/configuration/cli_options.py [hunk 2]
```diff
@@ -236,7 +239,7 @@ def __init__(self, *args, **kwargs):
         help='Specify the class name of the hyperopt loss function class (IHyperOptLoss). '
         'Different functions can generate completely different results, '
         'since the target for optimization is different. Built-in Hyperopt-loss-functions are: '
-        'DefaultHyperOptLoss, OnlyProfitHyperOptLoss, SharpeHyperOptLoss.'
+        'DefaultHyperOptLoss, OnlyProfitHyperOptLoss, SharpeHyperOptLoss '
```
Comment (hroff-1902) [inline freqtrade/configuration/cli_options.py]: `SharpeHyperOptLoss.'` (dot)

File: freqtrade/main.py [hunk 3]
```diff
@@ -33,16 +32,16 @@ def main(sysargv: List[str] = None) -> None:
...
+            return_code = args['func'](args)
         else:
-            # Load and run worker
-            worker = Worker(args)
-            worker.run()
+            # No subcommand was issued.
+            raise OperationalException(
+                "Usage of freqtrade requires a subcommand.\n"
+                "To use the previous behaviour, run freqtrade with `freqtrade trade [...]`.\n"
+                "To see a full list of options, please use `freqtrade --help`"
```
Comment (hroff-1902) [inline freqtrade/main.py]:
```
                "Usage of Freqtrade now requires a subcommand to be specified.\n"
                "To have the previous behavior (bot executing trades in live/dry-run modes, "
                "depending on the value of the `dry_run` setting in the config), run freqtrade as `freqtrade trade [options...]`.\n"
                "To see the full list of options available, please use `freqtrade --help` or `freqtrade command --help`."
```

**Comment by hroff-1902** [comment]
Copied from the Slack conversation:

* The codebase still contains 'HyperOpts' (in user_data). I thought we renamed already all of them in freqtrade and in the tests.
* I'd like to change the message printed when freqtrade is run without parameters. (see above)
* help string (printed with --help): I guess 'Common options' should be printed at the top, but there are some other inconsistencies in it, we'll be able to polish the helpstrings later. So just a comment, no changes required at this stage.

**Comment by xmatthias** [comment]
> The codebase still contains 'HyperOpts' (as 'OptS' in user_data). I thought we renamed already all of them in freqtrade and in the tests.

I thought so too - in #2388 ... 

but then again - to quote yourself from here: https://github.com/freqtrade/freqtrade/pull/2388#discussion_r336728315

> ... there are many 'hyperopts' left (where it's convenient and correct)

I would not rename the folder from `hyperopts` to `hyperopts` at this point - since this will be a pain point where hyperopt files will not be refound after the upgrade.

we'll need to rework the whole user_data stuff for pypi anyway (i got something in the works for that) ... so i'd like to leave that as is now.

**Comment by hroff-1902** [comment]
> > The codebase still contains 'HyperOpts' (as 'OptS' in user_data). I thought we renamed already all of them in freqtrade and in the tests.
> 
> I thought so too - in #2388 ...
> 
> but then again - to quote yourself from here: [#2388 (comment)](https://github.com/freqtrade/freqtrade/pull/2388#discussion_r336728315)
> 
> > ... there are many 'hyperopts' left (where it's convenient and correct)
> 
> I would not rename the folder from `hyperopts` to `hyperopts` at this point - since this will be a pain point where hyperopt files will not be refound after the upgrade.
> 
> we'll need to rework the whole user_data stuff for pypi anyway (i got something in the works for that) ... so i'd like to leave that as is now.

No need to change the name of the `hyperopts` dir.

I mean, a hyperopt class name there contains 'Opts' (SampleHyperOpts), which is _not_ convenient.

File: docs/bot-usage.md [hunk 4]
```diff
@@ -115,7 +131,7 @@ You can add the entry "user_data_dir" setting to your configuration, to always p
 Alternatively, pass in `--userdir` to every command.
 The bot will fail to start if the directory does not exist, but will create necessary subdirectories.
 
-This directory should contain your custom strategies, custom hyperopts and hyperopt loss functions, backtesting historical data (downloaded using either backtesting command or the download script) and plot outputs.
+This directory should contain your custom strategies, custom hyperopt and hyperopt loss functions, backtesting historical data (downloaded using either backtesting command or the download script) and plot outputs.
```
Comment (hroff-1902) [inline docs/bot-usage.md]: "custom hyperopts"

Comment (xmatthias) [reply docs/bot-usage.md]: aaaah yes, replaced one too much...

File: freqtrade/configuration/arguments.py (hunk 1 above)
Comment (hroff-1902) [reply freqtrade/configuration/arguments.py]: It's okay for now, let's leave it as it is, thanks.
👍

**Pull request merged**